### ISSUE TITLE
Feature request: user-defined collation sequences

### ISSUE THREAD

**Issue by xjfnet** [msallin/SQLiteCodeFirst#96]

**Comment by msallin** [comment]
Can you please provide details about how this feature should work?

**Comment by xjfnet** [comment]
sqlite Collation Sequence affect ORDER BY result;
the default Collation Sequence is BINARY
if the a column contains Chinese, this will get the wrong ORDER BY result;
so we need to do two things to get the right ORDER BY result:
1. register custom Collation Function PinYin
2. create table with column Collation Sequence specified PinYin

run this code with LINQPad

```cs
using Dapper;
using System.Data.SQLite;

void Main()
{
	var builder = new SQLiteConnectionStringBuilder();
	builder.DataSource = ":memory:";
	var db = new SQLiteConnection(builder.ToString());
	db.Open();

	/* this is the right order of the chinese string: 文, 中 */
	var list = new List<string>() { "中", "文" };
	list.OrderBy(x => x).Dump();

	/* create table without column collation sequence specified  */
	var cmd = db.CreateCommand();
	cmd.CommandType = CommandType.Text;
	cmd.CommandText = "CREATE TABLE [TableWithoutCollation] ( Name text )";
	cmd.ExecuteNonQuery();	

	db.Execute("INSERT INTO [TableWithoutCollation](Name) VALUES (@Name)", new[]
	{
		new { Name = "中" },
		new { Name = "文" } }
	);
	/* we get the wrong order: 中, 文 */
	db.Query("SELECT * FROM [TableWithoutCollation] ORDER BY [Name]").Dump();

	/* register collating function PinYin */
	SQLiteFunction.RegisterFunction(typeof(SQLiteCollation_PinYin));

	/* create table with column collation sequence specified  */
	cmd.CommandText = "CREATE TABLE [TableWithCollation] ( Name text COLLATE PinYin )";
	cmd.ExecuteNonQuery();

	db.Execute("INSERT INTO [TableWithCollation](Name) VALUES (@Name)", new[]
	{
		new { Name = "中" },
		new { Name = "文" } }
	);
	/* we get the right order: 文, 中 */
	db.Query("SELECT * FROM [TableWithCollation] ORDER BY [Name]").Dump();

	db.Close();
}

[SQLiteFunction(FuncType = FunctionType.Collation, Name = "PinYin")]
class SQLiteCollation_PinYin : SQLiteFunction
{
	public override int Compare(string x, string y)
	{
		return string.Compare(x, y);
	}
}
```

**Comment by msallin** [comment]
It is already possible to specify a built-in collation [CollateAttribute](https://github.com/msallin/SQLiteCodeFirst/blob/master/SQLite.CodeFirst/Public/Attributes/CollateAttribute.cs).

My idea is to extend the [CollationFunction enum](https://github.com/msallin/SQLiteCodeFirst/blob/master/SQLite.CodeFirst/Public/Attributes/CollationFunction.cs) and offer a property `FunctionName` on the `CollateAttribute`.

```
    public class Person
    {
        public int Id { get; set; }

        [MaxLength(50)]
        [Collate(CollationFunction.Custom, FunctionName = "PinYin")]
        public string FirstName { get; set; }
    }
```

You still have to add the SQLiteFunction by yourself.
Let me know what you think.

**Comment by msallin** [comment]
https://github.com/msallin/SQLiteCodeFirst/pull/105

**Comment by xjfnet** [comment]
i test #96, that works.

**Comment by msallin** [comment]
Thank you for creating the issue, the feature is included in the version 1.5.
To support this project you can
- star the repository;
- report bugs/request features by creating new issues;
- write code and create PRs;
- donate.

https://github.com/msallin/SQLiteCodeFirst/tree/v1.5.0.21#support-the-project-
